load-test-framework/src/main/java/com/google/pubsub/flic/cps/CPSPublishingTask.java

```diff
@@ -0,0 +1,145 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+package com.google.pubsub.flic.cps;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.flic.common.Utils;
+import com.google.pubsub.flic.kafka.KafkaConsumerTask;
+import com.google.pubsub.flic.processing.MessageProcessingHandler;
+import com.google.pubsub.flic.task.CPSTask;
+import com.google.pubsub.flic.task.TaskArgs;
+import com.google.pubsub.v1.PublishRequest;
+import com.google.pubsub.v1.PublishResponse;
+import com.google.pubsub.v1.PubsubMessage;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Runs a publishing task that sends messages to Cloud Pub/Sub using gRPC. */
+public class CPSPublishingTask extends CPSTask {
+  private static final Logger log = LoggerFactory.getLogger(KafkaConsumerTask.class.getName());
+
+  private CPSRoundRobinPublisher publisher;
+  private MessageProcessingHandler processingHandler;
+
+  public CPSPublishingTask(
+      TaskArgs args, CPSRoundRobinPublisher publisher, MessageProcessingHandler processingHandler) {
+    super(args);
+    this.publisher = publisher;
+    this.processingHandler = processingHandler;
+  }
+
+  /** Publishes messages to Cloud Pub/Sub in the most optimal way possible */
+  public void execute() throws Exception {
+    List<String> topics = args.getTopics();
+    String cpsProject = args.getCPSProject();
+    // The base message which will have a number appended to it.
+    String baseMessage = Utils.createMessage(args.getMessageSize(), 0);
+    PubsubMessage.Builder messageBuilder = PubsubMessage.newBuilder();
+    PublishRequest.Builder requestBuilder = PublishRequest.newBuilder();
+    // Warm up all threads in the thread pool.
+    callbackExecutor.prestartAllCoreThreads();
+    // Get the overall start time.
+    long start = System.currentTimeMillis();
+    while (messageNo.intValue() <= args.getNumMessages() && !failureFlag.get()) {
+      List<PubsubMessage> batch = new ArrayList<>();
+      // Keep track of the total # of bytes in this batch and the size of the batch.
+      AtomicLong batchBytes = new AtomicLong(0);
```

Comment:
Do batchBytes and counter need to be atomic? I see we use them later in a future, but by that point, we essentially hand ownership over to the future and don't make any more updates to the values.